src/renderers/shared/reconciler/instantiateReactComponent.js

```diff
@@ -83,6 +106,8 @@ function instantiateReactComponent(node, parentCompositeType) {
       // represenations. I.e. ART. Once those are updated to use the string
       // representation, we can drop this code path.
       instance = new element.type(element);
+    } else if (isStatelessComponentType(element.type)) {
```

Comment:
So, I don't see any option how to check function is stateless or module before `ReactCompositeComponentMixin.mountComponent`. That means we should merge `ReactCompositeComponentMixin` and `ReactStatelessComponentMixin` and turn on optimisations if `var inst = new Component(publicProps, publicContext);` returns `ReactElement`.

May be you have other better and simpler ideas?